From c0bc71f0634328745d0bbbf33bb1060807d0414c Mon Sep 17 00:00:00 2001
From: ElonH <elonhhuang@gmail.com>
Date: Sat, 28 Mar 2020 21:07:19 +0800
Subject: [PATCH] cpulimit-ng: missing sys/sysctl.h

--- a/src/cpulimit.c
+++ b/src/cpulimit.c
@@ -58,7 +58,7 @@
 #include <sys/stat.h>
 #include <sys/time.h>
 #include <sys/types.h>
-#include <sys/sysctl.h>
+#include <linux/sysctl.h>
 #include <sys/resource.h>
 #include <sys/wait.h>
 
